### PR TITLE
Initial Docker support

### PR DESCRIPTION
**Pull request by ypcs**
Run Docker inside Ubuntu VM (xenial) to allow easy migration to our new model.

This

 - removes existing Ansible roles that installed developer tools inside VM
 - installs Docker inside VM using new role
 - adds some helpers for adding backwards compatibility with existing seravo/wordpress checkouts

### CODE REVIEW

**Comment by ypcs** [comment]
Note: This will soon get couple patches, waiting for test build to finish. (E: patch merged)

File: roles/docker/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+docker_image: seravo/wordpress
```
Comment (ottok) [inline roles/docker/defaults/main.yml]: Maybe unnecessary? These are unlikely to change and the code is easier to read if there is less variables and these values could be embedded right where they are used.

File: roles/docker/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  apt:
+    pkg: "{{ item }}"
+    state: present
+    update_cache: yes
+  with_items:
+    - ca-certificates
+    - curl
+    - python-pip
+    - unattended-upgrades
+- name: add apt key for docker
```
Comment (ottok) [inline roles/docker/tasks/main.yml]: No space between sections makes the roles contents hard to read. Good practice would be to have space before every ' -name'

File: roles/docker/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    groups: docker
+    append: yes
+- name: create custom shell for jumping to docker
+  copy:
+    src: dockershell
+    dest: /bin/dockershell
+    mode: 0755
+    owner: root
+    group: root
+#- name: add dockershell as valid shell
```
Comment (ottok) [inline roles/docker/tasks/main.yml]: Do you really mean to keep this commented?

File: roles/docker/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      - '80:80'
+      - '443:443'
+      - '3306:3306'
+      - '2222:22'
+    restart_policy: always
+    state: stopped
+    env:
+      WP_USER: vagrant
+      WP_USER_PASS: vagrant
+#- name: change user shell
```
Comment (ottok) [inline roles/docker/tasks/main.yml]: Commented section here as well

**Review by ottok — CHANGES_REQUESTED** [review]
Looks very promising. Just a few tweaks and it will be perfect!

**Comment by ottok** [comment]
Built locally with 'vagrant up --provision' and 'vagrant package'.

Resuting box size was 1,5 GB. Old was less than 700 MB. Please take a look into slimming it down.

I added the local box to the local box registry with `vagrant box add --name seravo/wordpress-testi package.box'

Then in a clean seravo/wordpress project folder I modified Vagrantfile to use the local image name and ignore version requirements. Finally started with `vagrant up`.

During the startup there is a lot of "WARNING: Your Vagrantfile contains deprecated configuration.". I read the source, and I know why it says so, but I suggest that for now we just silently execute the wrapper and start to complain about it only in 6 months or so when we want to roll-out other upgrades to the Vagrantfile as well.

Once to box was up the address http://wordpress.local loaded correctly. However, https did not load and logs say `2018/05/04 02:15:32 [error] 18562#18562: *36 no "ssl_certificate" is defined in server listening on SSL port while SSL handshaking, client: 192.168.240.1, server: 0.0.0.0:443`

Also I noticed the script `wp-watch-logs` is missing. In general the list of wp-* seems lacking:
```
vagrant@wordpress:/$ wp-
wp-activate-git-hooks  wp-restart-nginx       wp-vagrant-activation  wp-vagrant-import-db
wp-generate-ssl        wp-use-asset-proxy     wp-vagrant-dump-db 
```

I also looks like the command `vagrant ssh` did not enter the inner Docker image but the host Virtual box image only and I was able to see all docker proxy processes etc. This is probably also why the wp-* were lacking.

File: roles/docker/defaults/main.yml (hunk 1 above)
Comment (ypcs) [reply roles/docker/defaults/main.yml]: Removed roles/docker/defaults & hard-coded these values.

File: roles/docker/tasks/main.yml (hunk 2 above)
Comment (ypcs) [reply roles/docker/tasks/main.yml]: Added space between all sections.

File: roles/docker/tasks/main.yml (hunk 3 above)
Comment (ypcs) [reply roles/docker/tasks/main.yml]: Removed.

File: roles/docker/tasks/main.yml (hunk 4 above)
Comment (ypcs) [reply roles/docker/tasks/main.yml]: Removed.

**Comment by ypcs** [comment]
Most of the size comes from our Docker image. While running wp-palvelu-vagrant image from this branch, unpacked size of Docker image is 1,7 gigabytes out of 3,0 gigabytes total.

However, I added two new tasks to slim down the VM by removing unnecessary clutter (around 300MB) from Ubuntu base image. Also one task to upgrade packages to latest versions. If we could remove linux-headers (required by virtualbox guest additions) we'd save another 100MB.

If we'd like to slim down the Vagrant image even further, we'd need to tweak the Docker image. I'll open a separate issue for this (https://github.com/Seravo/wp-palvelu-docker/issues/159).

In this branch, in addition to requests in review, also pushed fix to disable "please update" warnings.

**Comment by ypcs** [comment]
Splitted cleanup tasks to separate role. Also, now cleanup using `apt` module as much as possible, and instead of running `rm`/`mkdir` commands via shell module, use `file` module.

**Comment by ypcs** [comment]
I'll test/fix remaining issues (`vagrant ssh` doesn't enter docker, SSL certs not working) now.

**Comment by ypcs** [comment]
After updated cleanup role, shrinking Docker image etc, packed box size is now ~910MB. Some improvements could still be done, but I believe this is enough for now? Box includes new features compared to previous version, after all :)

Tested:

```shell
# build box locally & import it as seravo/wordpress (to vagrant)
make rebuild
make import

# clone wordpress repository (tested as using existing repository, as well as new checkout)
git clone https://github.com/seravo/wordpress
cd wordpress

# bring box up, provision it if not already done
vagrant up

# enter the docker environment
vagrant ssh

# confirm that we are inside docker
which docker # shouldn't return anything
```

**Comment by ypcs** [comment]
Ready for re-evaluating, ping @ottok

**Comment by ypcs** [comment]
Everything works correctly when using

```$ apt policy ansible
ansible:
  Installed: 2.5.2+dfsg-1
  Candidate: 2.5.2+dfsg-1
  Version table:
 *** 2.5.2+dfsg-1 500
        500 https://deb.debian.org/debian sid/main amd64 Packages
        100 /var/lib/dpkg/status
     2.5.1+dfsg-1 500
        500 https://deb.debian.org/debian buster/main amd64 Packages

$ apt policy vagrant
vagrant:
  Installed: 2.0.2+dfsg-3
  Candidate: 2.0.2+dfsg-3
  Version table:
 *** 2.0.2+dfsg-3 500
        500 https://deb.debian.org/debian buster/main amd64 Packages
        500 https://deb.debian.org/debian sid/main amd64 Packages
        100 /var/lib/dpkg/status

$ apt policy virtualbox-5.2 
virtualbox-5.2:
  Installed: 5.2.10-122088~Debian~stretch
  Candidate: 5.2.10-122088~Debian~stretch
  Version table:
 *** 5.2.10-122088~Debian~stretch 100
        100 /var/lib/dpkg/status
```

**Comment by l3ku** [comment]
Built the vagrant image locally similarily as @ottok did, after modifying the Vagrantfile in Seravo/wordpress an running `vagrant up`, something goes wrong:
```
==> wordpress-box: Setting hostname...
==> wordpress-box: Configuring and enabling network interfaces...
==> wordpress-box: Mounting shared folders...
    wordpress-box: /data/wordpress => /Users/Leo/seravo/wordpress
==> wordpress-box: Running provisioner: shell...
    wordpress-box: Running: inline script
==> wordpress-box: Running provisioner: shell...
    wordpress-box: Running: inline script
==> wordpress-box: sudo: /usr/local/sbin/create-dirs: command not found
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```
Any advice on where to start looking for the problem?

**Comment by ypcs** [comment]
Fix for missing create-dirs added ^.

**Comment by l3ku** [comment]
The `vagrant up` command now gets past the "Running: inline script" phase, but fails shortly afterwards:
```
==> wordpress-box: Setting hostname...
==> wordpress-box: Configuring and enabling network interfaces...
==> wordpress-box: Mounting shared folders...
    wordpress-box: /data/wordpress => /Users/Leo/seravo/wordpress
==> wordpress-box: Running provisioner: shell...
    wordpress-box: Running: inline script
==> wordpress-box: Running provisioner: shell...
    wordpress-box: Running: inline script
==> wordpress-box: dockershell /usr/local/bin/wp-vagrant-activation
==> wordpress-box: seravo_wordpress
==> wordpress-box: Waiting for environment...
==> wordpress-box: /usr/local/bin/wp-vagrant-activation:21:in `<main>': ERROR: config.yml was not found. Please provide the needed information for your box in config.yml. (RuntimeError)
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```
This is very strange, as the error is caused by /data/wordpress/config.yml not existing in the Docker image, although it exists in the vagrant image:
```
INSIDE DOCKER CONTAINER
root@vagrant:~# ls -a /data/wordpress/
.  ..
root@vagrant:~#
```
```
INSIDE VAGRANT
vagrant@wordpress:/data/wordpress$ ls /data/wordpress/
composer.json  config-sample.yml  gulpfile.js  LICENSE  package.json  scripts  Vagrantfile
composer.lock  config.yml         htdocs       nginx    README.md     tests    vendor
vagrant@wordpress:/data/wordpress$
```
Docker inspect mount data:
```
"Mounts": [
            {
                "Type": "bind",
                "Source": "/data",
                "Destination": "/data",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
```
The image was built from the state after b55bc6c6ca95f5b1826b8ce6543b2a11ec99038f.

**Comment by ypcs** [comment]
I was able to confirm ^, that Vagrant box built on OS X from b55bc6c fails to start. Now re-rebuilding from same commit on Linux.

**Comment by ypcs** [comment]
Mount was empty because Docker mounted it's volume before Virtualbox did it's mount. So, empty /data got mounted into docker/data, and host VM had real contents @ /data.

These issues are fixed now, squashed pushes & just waiting for last test build to finish before merging.

**Pull request merged**